### ISSUE TITLE
[Feature] Use context of different size in encoder with cross attention

### ISSUE THREAD

**Issue by Aceticia** [lucidrains/x-transformers#302]
Hi! I'm back with another simple feature request. As always, I'm always happy to contribute myself if you are busy with other stuff. 

**Why this feature**: I'm recently experimenting with MAE pretrained models with cross attention decoders rather than self attention. You might know that since the decoder has a smaller size than the encoder, we need to reduce the dimension in between the encoder and decoder which causes a lot of lost information when masking rate is high. If we use cross attention, we can instead keep the encoder dimension and simply use a bigger key and value projection matrix during the cross attention.

Currently, the `Attention` class already supports a `dim_context` input, but there is no way to specify it in `AttentionLayers`'s init call. (Just passing `attn_dim_context=xxx` would cause all layers to assume the same context size, which we don't want for self-attention layers) We can just add a new `dim_context` argument to `AttentionLayers` and only pass it into `Attention` when layer type is `c`. What do you think? I haven't found any other hacks to achieve this so far, but if you do know please also let me know!

**Comment by lucidrains** [comment]
hey Chris! yes I can add this, and in fact, I'll try to add this while out and about with doggo, using mobile tethering

let me find a good rest spot..

**Comment by lucidrains** [comment]
@Aceticia oh, actually i think it is supported already, just not documented

does this work for you?

```python
import torch
from x_transformers import Decoder


x = torch.randn((1, 64, 256))
mask = torch.ones((1, 64), dtype=torch.bool)
context = torch.randn((1, 128, 253))
context_mask = torch.ones((1, 128), dtype=torch.bool)
model = Decoder(dim=256, depth=4, heads=4, alibi_pos_bias=True, cross_attend=True, cross_attn_dim_context = 253)

y = model(x=x, mask=mask, context=context, context_mask=context_mask)
```

**Comment by Aceticia** [comment]
Ah you are right. I completely missed these. Thanks a lot!! I'll close this then.

**Comment by Aceticia** [comment]
@lucidrains A few quick runs seem to suggest the model is indeed doing better at the pretraining task, but at the cost of a bit more parameters and a slightly smaller batch size. I'm not sure how this will impact downstream tasks - will let you know if I find anything interesting here!
